### PR TITLE
enable apm test agent service naming checks

### PR DESCRIPTION
**Pull request by wconti27**
This PR enables new APM Test Agent peer service and service naming checks. The checks required that the Span Attribute Schema version and expected service name be included in the trace header payload to the Test Agent. The CI changes add a new test step to query the APM Test Agent for Trace Check results (# of skipped, passed and failed checks) by enabled check name for Plugin tests. 

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

**Review by jbertran — COMMENTED** [review]
The current [failure](https://github.com/DataDog/dd-trace-js/actions/runs/5543549394/jobs/10119732689?pr=3353) on `redis` looks like the running `DD_TRACE_SPAN_ATTRIBUTE_SCHEMA` env var is `v0`, but the test agent receives the `v1` header 🤔 

Is it possible that we're overwriting existing configuration?

File: packages/dd-trace/test/plugins/agent.js [hunk 1]
```diff
@@ -36,19 +36,44 @@ function ciVisRequestHandler (request, response) {
...
-  // to send with trace to final agent destination
-  if (ddEnvVars.length > 0) {
-    headers['X-Datadog-Trace-Env-Variables'] = ddEnvVars.join(',')
-  }
+  return new Promise((resolve, reject) => {
+    // get all environment variables that start with "DD_"
+    const ddEnvVars = new Map(
+      Object.entries(process.env)
+        .filter(([key]) => key.startsWith('DD_'))
+        .map(([key, value]) => [key, value])
```
[the remark this answers is not in the document]
Comment (wconti27) [reply packages/dd-trace/test/plugins/agent.js]: no, removed it

File: packages/dd-trace/test/plugins/agent.js [hunk 2]
```diff
@@ -36,19 +36,44 @@ function ciVisRequestHandler (request, response) {
...
+    )
+    if (global.testAgentServiceName) {
+      ddEnvVars.set('DD_SERVICE', global.testAgentServiceName)
+    }
+    if (global.schemaVersionName) {
+      ddEnvVars.set('DD_TRACE_SPAN_ATTRIBUTE_SCHEMA', global.schemaVersionName)
+    }
+
+    for (let i = 0; i < plugins.length; i++) {
```
[the remark this answers is not in the document]
Comment (wconti27) [reply packages/dd-trace/test/plugins/agent.js]: done

**Review by jbertran — APPROVED** [review]

File: packages/dd-trace/test/setup/mocha.js [hunk 3]
```diff
@@ -74,10 +76,20 @@ function withNamingSchema (
           })
         })
 
+        beforeEach(async () => {
+          await new Promise(resolve => setTimeout(resolve, 30))
```
Comment (tlhunter) [inline packages/dd-trace/test/setup/mocha.js:80]: What's this about?

File: packages/dd-trace/test/setup/mocha.js [hunk 4]
```diff
@@ -113,10 +125,19 @@ function withNamingSchema (
         })
       })
 
+      beforeEach(async () => {
+        await new Promise(resolve => setTimeout(resolve, 30))
```
Comment (tlhunter) [inline packages/dd-trace/test/setup/mocha.js:129]: Between these two sleep statements it appears that we're slowing down every `withNamingSchema` test by 60ms?

File: packages/dd-trace/test/setup/mocha.js (hunk 3 above)
Comment (wconti27) [reply packages/dd-trace/test/setup/mocha.js:80]: I was encountering a data race where the mock agent was still handling traces from previous test cases while this test case was being started.

Comment (tlhunter) [reply packages/dd-trace/test/setup/mocha.js:80]: We'll need to isolate the cause of these race conditions instead of adding the slowdown. Maybe some day the CI servers will get slower and we'll get confusing errors that we don't know how to debug.

Is it specific tests that fail? Or is it random? It could be that a subset of tests are marked as complete even though they're still doing some background tasks.

**Review by tlhunter — CHANGES_REQUESTED** [review]

File: packages/dd-trace/test/setup/mocha.js [hunk 5]
```diff
@@ -17,6 +17,9 @@ global.withVersions = withVersions
 global.withExports = withExports
 global.withNamingSchema = withNamingSchema
 global.withPeerService = withPeerService
+global.testAgentServiceName = null
```
Comment (tlhunter) [inline packages/dd-trace/test/setup/mocha.js:20]: The existing global stateless test helper functions are used across a wide number of tests.

It looks like these new globals are very stateful and were added only for a single test file, `packages/dd-trace/test/plugins/agent.js`? It also looks like a pretty tight coupling now between mocha.js and agent.js.

A better way to access and mutate these values would be to stick them into another JavaScript module via exports. This makes them singleton values but no longer with the side effects of being globals. That said, I think there would still be a tighter-than-needed coupling between mocha.js and agent.js.

Should we hop in a meeting to figure out the best course of action?

**Review by tlhunter — CHANGES_REQUESTED** [review]

**Review by jbertran — DISMISSED** [review]
LGTM!

File: .github/actions/testagent/logs/action.yml [hunk 6]
```diff
@@ -15,3 +15,17 @@ runs:
           docker-compose logs testagent
         fi
       shell: bash
+    - uses: actions/checkout@v2
+    - run: |
+        RESPONSE=$(curl -s -w "\n%{http_code}" -o response.txt http://127.0.0.1:9126/test/trace_check/failures)
```
Comment (rochdev) [inline .github/actions/testagent/logs/action.yml:20]: Should this and other scripts be moved to JS files in the `scripts` folder? This would make it easier to test them locally.

File: .github/workflows/plugins.yml [hunk 7]
```diff
@@ -46,6 +46,8 @@ jobs:
       - run: yarn test:plugins:upstream
       - if: always()
         uses: ./.github/actions/testagent/logs
+      - if: always()
+        uses: ./.github/actions/testagent/results
```
Comment (rochdev) [inline .github/workflows/plugins.yml:50]: Why does this need to be added everywhere?

File: packages/dd-trace/test/plugins/agent.js [hunk 8]
```diff
@@ -35,59 +43,154 @@ function ciVisRequestHandler (request, response) {
...
+      // call original method
+      stubs.originalMethods._sendPayload.call(thisValue, data, count, done)
+    })
+  } else {
+    stubs.originalMethods._sendPayload.call(thisValue, data, count, done)
+  }
+}
+
+// create stub on the startSpan method to inject schema version and other tags
+const startSpanMock = function (name, { childOf, kind, meta, metrics, service, resource, type } = {}, enter = true) {
```
Comment (rochdev) [inline packages/dd-trace/test/plugins/agent.js:63]: Why is all this new code directly in the agent? Right now it doesn't really do expectations. Could the expectations be moved out?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: .github/workflows/plugins.yml (hunk 7 above)
Comment (rochdev) [reply .github/workflows/plugins.yml:50]: But then if that's always needed, why can't it be added to the existing `log` step?

Comment (wconti27) [reply .github/workflows/plugins.yml]: I was initially concerned with readability and wanted it to be clear to devs that there's both test agent results and logs. I just updated the PR to just use a single step for both logs and results (with better naming).

File: packages/dd-trace/test/plugins/agent.js [hunk 9]
```diff
@@ -35,59 +36,66 @@ function ciVisRequestHandler (request, response) {
...
-
-function handleTraceRequest (req, res, sendToTestAgent) {
+async function handleTraceRequest (req, res, sendToTestAgent) {
+  res.status(200).send({ rate_by_service: { 'service:,env:': 1 } })
   // handles the received trace request and sends trace to Test Agent if bool enabled.
   if (sendToTestAgent) {
     const testAgentUrl = process.env.DD_TEST_AGENT_URL || 'http://127.0.0.1:9126'
+    let schemaVersion
+    let expectedServiceName
+    let traceEnvHeader = ''
```
Comment (rochdev) [inline packages/dd-trace/test/plugins/agent.js:46]: Why is this needed? If we need to set it manually, then we're not really testing a real scenario.

**Comment by tlhunter** [comment]
@wconti27 is this PR still something that you'd like to land?
